// @vkontakte/vkjs/lib/datetime/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    SECONDS_IN_THE_DAY: function() {
        return _date.SECONDS_IN_THE_DAY;
    },
    addDays: function() {
        return _date.addDays;
    },
    convertDateToInputFormat: function() {
        return _date.convertDateToInputFormat;
    },
    createDateFromUnixTimestamp: function() {
        return _date.createDateFromUnixTimestamp;
    },
    formatDuration: function() {
        return _duration.formatDuration;
    },
    getBeginningOfDay: function() {
        return _date.getBeginningOfDay;
    },
    getLastDayOfMonth: function() {
        return _date.getLastDayOfMonth;
    },
    getStartOfWeek: function() {
        return _date.getStartOfWeek;
    },
    getUnixTimestampFromDate: function() {
        return _date.getUnixTimestampFromDate;
    },
    isDateToday: function() {
        return _isDateToday.isDateToday;
    },
    isDateTomorrow: function() {
        return _date.isDateTomorrow;
    },
    isDateYesterday: function() {
        return _date.isDateYesterday;
    },
    isLeapYear: function() {
        return _date.isLeapYear;
    },
    isSameDate: function() {
        return _isSameDate.isSameDate;
    }
});
var _date = require("../datetime/date");
var _isDateToday = require("./isDateToday");
var _isSameDate = require("./isSameDate");
var _duration = require("./duration");
